sympy/liealgebras/cartan_base.py

```
@@ -0,0 +1,286 @@
+        """
+        i = i - 1
+        return self.omega_matrix().row(i)
+
+    def fundamental_weights(self):
+        """
+        Returns a list of the fundamental weights in the orthogonal basis
+        """
+        return [self.fundamental_weight(i) for i in range(self.omega_matrix().rows)]
+
+    def rootsystem(self):
+        """Returns the root system of the group ordered from
+        highest root to lowest. The roots are found by reflecting
+        each simple root about its hyperplane, repeating this procedure
+        on each subsequent root generated until no more are found.
+        The roots are then weighed and sorted according to weight.
+
+        Note: This is a costly calculation for groups with
+        rank > 6, most notably E7 and E8
+
+        Examples
+        ========
+        >>> from sympy.liealgebras import CartanType
+        >>> CartanType("A2").rootsystem()
+        [Matrix([[1, 0, -1]]),
+        Matrix([[1, -1, 0]]),
+        Matrix([[0, 1, -1]]),
+        Matrix([[0, 0, 0]]),
+        Matrix([[0, 0, 0]]),
+        Matrix([[0, -1, 1]]),
+        Matrix([[-1, 1, 0]]),
+        Matrix([[-1, 0, 1]])]
+        """
+        s_r = self.simple_roots()
+        rank = self.rank()
+
+        orbits = set()
+        for i in s_r:
+            for r in self.orbit(i):
+                orbits.add(tuple(r))
+
+        zero_roots = [Matrix([0] * rank).T] * rank
+
+        orbits = [Matrix([[*i]]) * self.cocartan_matrix().T for i in orbits] + zero_roots
+
+        # sort roots by their weights
+        sorbits = sorted(orbits, key = lambda x: -self.root_level(x, 'alpha'))
+
+        # rotate back to the orthogonal basis for consistency
+        omega_matrix = self.omega_matrix()
+        return [x * omega_matrix for x in sorbits]
+
+    def roots(self):
+        """Returns the number of total roots in the algebra"""
+        warnings.warn("Do not call this method directly from the base class.")
```

It's better to raise an error (e.g. NotImplementedError) than warn if this is not supposed to be called.